apps/grader-host/src/index.ts

```diff
@@ -188,24 +203,77 @@ async function handleJob(job: GradingJobMessage) {
 
-    const initResults = await Promise.all([
+    const initResults = await Promise.allSettled([
       reportReceived(context, receivedTime),
+      // It's possible that this could fail for normal reasons, e.g. a Docker
+      // image not existing in our cache. We'll allow this error to bubble up
+      // but we won't report it to Sentry.
       initDocker(context),
-      initFiles(context),
+      initFiles(context).catch((err) => {
+        Sentry.captureException(err);
+        throw err;
+      }),
     ]);
 
-    const results = await runJob(context, receivedTime, initResults[2].tempDir);
+    const results = await run(async () => {
+      const initSucceeded = initResults.every((result) => result.status === 'fulfilled');
+      if (initSucceeded && initResults[2].status === 'fulfilled') {
+        const tempDir = initResults[2].value;
+        return await runJob(context, receivedTime, tempDir?.tempDir);
+      } else {
+        const message = run(() => {
+          const dockerInitSucceeded = initResults[1].status === 'fulfilled';
+          const fileInitSucceeded = initResults[2].status === 'fulfilled';
+
+          const messages: string[] = [];
+          if (!dockerInitSucceeded) {
+            messages.push(`Could not pull Docker image ${job.image}.`);
+          }
+          if (!fileInitSucceeded) {
+            messages.push('Could not initialize files for grading.');
+          }
+
+          // This should never occur, as `reportReceived` should never reject. But,
+          // just in case, we'll return a generic error message.
+          if (messages.length === 0) {
+            return 'An unknown error occurred.';
+          }
 
-    logger.info(`Job ${job.jobId} completed with results:`, results);
+          return messages.join('\n');
+        });
 
-    await Promise.all([
-      uploadResults(context, results),
-      uploadArchive(context, initResults[2].tempDir),
-    ]);
+        return {
+          job_id: job.jobId,
+          received_time: receivedTime,
+          start_time: new Date(),
+          end_time: new Date(),
+          succeeded: false,
+          message,
+        };
+      }
+    });
+
+    const tempDir = initResults[2].status === 'fulfilled' ? initResults[2].value : undefined;
 
-    logger.info('Removing temporary directories');
-    await initResults[2].tempDirCleanup();
-    logger.info('Successfully removed temporary directories');
+    logger.info(`Job ${job.jobId} completed with results:`, results);
+
+    try {
+      await Promise.all([
+        uploadResults(context, results),
+        // Only attempt to upload the archive if the directory was created.
+        tempDir ? uploadArchive(context, tempDir.tempDir) : Promise.resolve(),
+      ]);
+
+      logger.info('Removing temporary directories');
+      await tempDir?.tempDirCleanup();
+      logger.info('Successfully removed temporary directories');
+    } catch (err) {
+      Sentry.captureException(err);
+      throw err;
```

Comment:
I think there's still room to improve error handling a bit. It's not visible on this line, but higher up, we somewhat silently swallow errors from the execution process:

https://github.com/PrairieLearn/PrairieLearn/blob/e3eafc75bd2dc578113d8acc6c380aa9ddf21151/apps/grader-host/src/index.ts#L133

So, this PR still doesn't help if `uploadResults` fails, as the job will be deleted from SQS and won't be re-attempted. I'm wondering if we should change the top-level bit to allow the error to propagate up to the `receiveFromQueue` callback (in which case the message won't be removed from SQS) so that if `uploadResults` fails, the job as a whole will be retried via SQS?